### PR TITLE
Explorer Intern Project: Clone Locally and App Settings

### PR DESCRIPTION
**Pull request by faditawfik21**
ADDED FILES: 

`src/commands/dockersupport/askDockerStep.ts` - Prompt pop up UI at the top of VS Code that asks if the user would like to use Docker with their project with yes/no options only. Returns string 
GetImage

`src/commands/dockersupport/localDockerSupport.ts` - function that checks if certain conditions are met to open the App Function locally with Docker. If Docker is not installed, it points users to the documentation. Returns boolean 
GetImage (1)

`src/commands/dockersupport/validateDockerInstalled.ts` - checks the user's system if Docker is installed by running a command in the command line. Returns boolean 

`src/downloadAzureproject/cloneProjectLocally.ts` - The function call of the new menu item button "Clone Project Locally…" to trigger the whole download flow 

`src/commands/appSettings/filterDownloadAppSettings.ts` - This file contains the new logic behind downloading app settings. Using the user's list of app settings, a quickpick item is created with any app setting containing secrets unpicked by default. The quickPick is shown to the user and based on the user's selection, the value of any unpicked app setting will not be downloaded and rather, set to `_REDACTED_`. Additionally, a list of the app settings chosen to not be downloaded will be stored under the `SettingsToIgnoreOnDeployment` in the local settings file. The output is then logged. 
sc of UI 

`src/commands/appSettings/filterUploadAppSettings.ts` - This file contains the new logic behind uploading app settings. If an app setting was stored in the `SettingsToIgnoreOnDeployment` field then it will be automatically ignored during the upload flow and its portal value will not be modified. If a setting is not in the `SettingsToIgnoreOnDeployment` field but is modified, then a pop-up will appear prompting the user to confirm the overwrite like in the original upload app settings flow. Once done, the corresponding log will be shown. 
sc of upload pop up
sc of upload log

---------------------------- 

EDITED FILES: 

`src/tree/SlotTreeItemBase.ts` - added a function called getApplicationLanguage() which goes through the app function settings and gets its runtime. Returns string 

`package.json` - Edited to add menu item placement and to update version of extract 

`package.nls.json` - Edited to add menu item, specifically to add the words that go on the button 
GetImage (2)

`src/commands/registerCommands.ts` - Edited to add menu item, specifically to relate the button to a function call 

`src/downloadAzureproject/setUpProjectFolder.ts` - Edited to handle calls from the extension side and to include call for docker prompt 

`package-lock.json` - Edited to update version of extract 

`src/templates/script/ScriptTemplateProvider.ts` - Edited to update extract version, specifically adapting to the fact that the new version of extract no longer allows callbacks and forces you to promises instead 

`src/commands/funcConfig/local.settings.ts` - A "SettingsToIgnoreOnDeployment" string array was added to the `ILocalSettingsJson` interface in order to store any app settings that should be not be uploaded. 
sc of SettingsToIgnoreOnDeployment

`src/commands/appSettings/downloadAppSettings.ts` - The `SettingsToIgnoreOnDeployment`field is initialized if found empty and a function call to `filterDownloadAppSettings.ts` is made. 

`src/commands/appSettings/uploadAppSettings.ts` - The `SettingsToIgnoreOnDeployment` field is initialized if found empty and a function call to `filterUploadAppSettings.ts` is made. 

---------------------------- 

DELETED FILES: 

`src/commands/confirmOverwriteSettings.ts` - No longer needed as it was replaced with `filterDownloadAppSettings.ts` and `filterUploadAppSettings.ts`.

### CODE REVIEW

File: .vscode/tasks.json [hunk 1]
```diff
@@ -18,6 +18,32 @@
             "type": "npm",
             "script": "lint",
             "problemMatcher": "$eslint-stylish"
+        },
+        {
```
Comment (refortie) [inline .vscode/tasks.json:22]: Confirming you intended this file to be added?

File: host.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (refortie) [inline host.json:1]: Confirming you intended this file to be added?

File: proxies.json [hunk 3]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (refortie) [inline proxies.json:1]: Confirming you intended this file to be added?

File: src/commands/appSettings/filterUploadAppSettings.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                    destinationSettings[key] = sourceSettings[key];
+                } else {
+                    userIgnoredKeys.push(key);
+                }
+            }
+        }
+        else {
+            securitySettingsIgnored.push(key);
+        }
+
```
Comment (refortie) [inline src/commands/appSettings/filterUploadAppSettings.ts:49]: Nit: extra newline

File: src/commands/appSettings/downloadAppSettings.ts [hunk 5]
```diff
@@ -62,15 +69,17 @@ export async function downloadAppSettingsInternal(context: IActionContext, clien
...
-            const doc: vscode.TextDocument = await vscode.workspace.openTextDocument(localSettingsUri);
-            await vscode.window.showTextDocument(doc);
-        } else if (result === viewOutput) {
-            ext.outputChannel.show();
-        }
-    });
+    if (showMessage) {
+        ext.outputChannel.appendLog(localize('downloadedSettings', 'Successfully downloaded settings.'), { resourceName: client.fullName });
+        const openFile: string = localize('openFile', 'Open File');
+        // don't wait
```
Comment (refortie) [inline src/commands/appSettings/downloadAppSettings.ts:75]: Do you need this comment?

File: src/commands/appSettings/filterDownloadAppSettings.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    for (const element of Object.keys(sourceSettings)) {
+        options.push({
+            label: element
+        });
+    }
+
+    const result = await context.ui.showQuickPick(options, {
+        placeHolder: 'Select the app settings you would like to download:',
+        isPickSelected: (item) => { return !listOfSettingsToIgnore.includes(item.label) }
+        , canPickMany: true
```
Comment (refortie) [inline src/commands/appSettings/filterDownloadAppSettings.ts:30]: I thought this was fixed in a previous PR, but can you put this on the previous line

File: src/commands/deploy/deploy.ts [hunk 7]
```diff
@@ -87,7 +87,7 @@ async function deploy(actionContext: IActionContext, arg1: vscode.Uri | string |
         await verifyAppSettings(context, node, projectPath, version, language, { doRemoteBuild, isConsumption });
     }
 
-    await node.runWithTemporaryDescription(
+    await node.runWithTemporaryDescription( // issues temporary description while call back is being run
```
Comment (refortie) [inline src/commands/deploy/deploy.ts:90]: Do we need this comment?

File: src/commands/executeFunction.ts [hunk 8]
```diff
@@ -40,13 +40,15 @@ export async function executeFunction(context: IActionContext, node?: FunctionTr
         }
 
         const data: string = await context.ui.showInputBox({ prompt, value, stepName: 'requestBody' });
+
```
Comment (refortie) [inline src/commands/executeFunction.ts:43]: Perhaps revert this file?

File: src/commands/initProjectForVSCode/initProjectForVSCode.ts [hunk 9]
```diff
@@ -42,6 +42,7 @@ export async function initProjectForVSCode(context: IActionContext, fsPath?: str
         return;
     }
 
+    // get settings, language, version for the project
```
Comment (refortie) [inline src/commands/initProjectForVSCode/initProjectForVSCode.ts:45]: Not sure we need this comment

File: src/commands/registerCommands.ts [hunk 10]
```diff
@@ -53,6 +54,7 @@ import { swapSlot } from './swapSlot';
 import { disableFunction, enableFunction } from './updateDisabledState';
 import { viewProperties } from './viewProperties';
 
+
```
Comment (refortie) [inline src/commands/registerCommands.ts:57]: Extra new line

File: src/commands/registerCommands.ts [hunk 11]
```diff
@@ -90,7 +92,7 @@ export function registerCommands(): void {
     registerCommand('azureFunctions.openInPortal', openInPortal);
     registerCommand('azureFunctions.pickProcess', pickFuncProcess);
     registerSiteCommand('azureFunctions.redeploy', redeployDeployment);
-    registerCommand('azureFunctions.refresh', async (context: IActionContext, node?: AzureTreeItem) => await ext.tree.refresh(context, node));
+    registerCommand('azureFunctions.refresh', async (context: IActionContext, node?: AzureTreeItem) => await ext.tree.refresh(context, node)); // menu item 5
```
Comment (refortie) [inline src/commands/registerCommands.ts:95]: Do you need this comment?

**Review by refortie — CHANGES_REQUESTED** [review]

File: src/commands/appSettings/filterUploadAppSettings.ts (hunk 4 above)
Comment (faditawfik21) [reply src/commands/appSettings/filterUploadAppSettings.ts:49]: removed

File: src/commands/appSettings/downloadAppSettings.ts (hunk 5 above)
Comment (faditawfik21) [reply src/commands/appSettings/downloadAppSettings.ts:75]: removed

File: src/commands/deploy/deploy.ts (hunk 7 above)
Comment (faditawfik21) [reply src/commands/deploy/deploy.ts:90]: removed

File: .vscode/tasks.json (hunk 1 above)
Comment (faditawfik21) [reply .vscode/tasks.json:22]: This file is not needed so I will delete it.

File: host.json (hunk 2 above)
Comment (faditawfik21) [reply host.json:1]: This file is not needed so I will delete it.

File: proxies.json (hunk 3 above)
Comment (faditawfik21) [reply proxies.json:1]: This file is not needed so I will delete it.

File: src/commands/appSettings/filterUploadAppSettings.ts (hunk 4 above)
Comment (faditawfik21) [reply src/commands/appSettings/filterUploadAppSettings.ts:49]: Deleted.

File: src/commands/appSettings/downloadAppSettings.ts (hunk 5 above)
Comment (faditawfik21) [reply src/commands/appSettings/downloadAppSettings.ts:75]: No need, I will delete it.

File: src/commands/deploy/deploy.ts (hunk 7 above)
Comment (faditawfik21) [reply src/commands/deploy/deploy.ts:90]: No need, I will delete it.

File: src/commands/executeFunction.ts (hunk 8 above)
Comment (faditawfik21) [reply src/commands/executeFunction.ts:43]: Sure.

File: src/commands/initProjectForVSCode/initProjectForVSCode.ts (hunk 9 above)
Comment (faditawfik21) [reply src/commands/initProjectForVSCode/initProjectForVSCode.ts:45]: I will delete it.

File: src/commands/registerCommands.ts (hunk 10 above)
Comment (faditawfik21) [reply src/commands/registerCommands.ts:57]: Deleted.

File: src/commands/registerCommands.ts (hunk 11 above)
Comment (faditawfik21) [reply src/commands/registerCommands.ts:95]: No need, I will delete it.

File: src/commands/appSettings/filterDownloadAppSettings.ts (hunk 6 above)
Comment (faditawfik21) [reply src/commands/appSettings/filterDownloadAppSettings.ts]: Fixed.

File: .vscode/tasks.json [hunk 12]
```diff
@@ -1,23 +0,0 @@
-{
```
Comment (nertim) [inline .vscode/tasks.json:1]: should this get readded?

Comment (faditawfik21) [reply .vscode/tasks.json:1]: It should. I submitted a PR to update our internship master branch with `tasks.json` and waiting for it to be approved.

**Pull request closed without merging**